crates/tabby-scheduler/src/incremental.rs

```diff
@@ -0,0 +1,149 @@
+use std::{
+    collections::HashSet,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use anyhow::Result;
+use kv::{Bucket, Config, Json, Store};
+use tabby_common::{config::RepositoryConfig, SourceFile};
+
+use crate::{
+    code::CodeIntelligence,
+    dataset::{create_source_file, RepositoryExt},
+    path,
+};
+
+pub struct IncrementalRepositoryStore {
+    store: Store,
+}
+
+const VERSIONS_BUCKET: &str = "versions";
+const DATASET_BUCKET: &str = "dataset";
+
+fn cmd_stdout(pwd: &Path, cmd: &str, args: &[&str]) -> Result<String> {
+    Ok(String::from_utf8(
+        Command::new(cmd)
+            .current_dir(pwd)
+            .args(args)
+            .spawn()?
+            .wait_with_output()?
+            .stdout,
+    )?)
+}
+
+fn get_git_commit(path: &Path) -> Result<String> {
+    cmd_stdout(path, "git", &["rev-parse", "HEAD"])
+}
+
+fn get_changed_files(path: &Path, since_commit: String) -> Result<Vec<String>> {
+    cmd_stdout(path, "git", &["diff", "--name-only", &since_commit])
+        .map(|s| s.lines().map(|line| line.to_owned()).collect())
+}
+
+fn repository_file_key(base_dir: &str, path: &str) -> String {
+    format!("{base_dir}:{path}")
+}
+
+fn parse_key_base_dir(key: &str) -> &str {
+    key.split_once(':').unwrap().0
+}
+
+impl IncrementalRepositoryStore {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            store: Store::new(Config::new(path::incremental_repository_store()))?,
+        })
+    }
+
+    fn repository_versions_bucket(&self) -> Bucket<String, String> {
+        self.store.bucket(Some(VERSIONS_BUCKET)).unwrap()
+    }
+
+    fn dataset_bucket(&self) -> Bucket<String, Json<SourceFile>> {
+        self.store.bucket(Some(DATASET_BUCKET)).unwrap()
+    }
+
+    fn set_last_commit(&self, repo_path: String, commit_hash: String) -> Result<()> {
+        self.repository_versions_bucket()
+            .set(&repo_path, &commit_hash)?;
+        Ok(())
+    }
+
+    fn get_last_commit(&self, repo_path: String) -> Result<Option<String>> {
+        Ok(self.repository_versions_bucket().get(&repo_path)?)
+    }
+
+    pub fn update_repository(&self, repository: &RepositoryConfig) -> Result<()> {
```

Comment:
consider use transaction to ensure update to each repo is atomic:

https://docs.rs/kv/latest/kv/struct.Transaction.html